pkg/gui/controllers/helpers/fixup_helper.go

```diff
@@ -0,0 +1,175 @@
+package helpers
+
+import (
+	"regexp"
+	"strings"
+	"sync"
+
+	"github.com/jesseduffield/generics/set"
+	"github.com/jesseduffield/lazygit/pkg/commands/models"
+	"github.com/jesseduffield/lazygit/pkg/gui/types"
+	"github.com/jesseduffield/lazygit/pkg/utils"
+	"github.com/samber/lo"
+)
+
+type FixupHelper struct {
+	c *HelperCommon
+}
+
+func NewFixupHelper(
+	c *HelperCommon,
+) *FixupHelper {
+	return &FixupHelper{
+		c: c,
+	}
+}
+
+type deletedLineInfo struct {
+	filename     string
+	startLineIdx int
+	numLines     int
+}
+
+func (self *FixupHelper) HandleFindBaseCommitForFixupPress() error {
+	diff, useIndex, err := self.getDiff()
+	if err != nil {
+		return err
+	}
+	if diff == "" {
+		return self.c.ErrorMsg("No changes to commit")
+	}
+
+	deletedLineInfos := self.parseDiff(diff)
+	if len(deletedLineInfos) == 0 {
+		return self.c.ErrorMsg("No deleted lines in diff")
+	}
+
+	shas := self.blameDeletedLines(deletedLineInfos)
+
+	if len(shas) == 0 {
+		// This should never happen
+		return self.c.ErrorMsg("No base commits found")
+	}
+	if len(shas) > 1 {
+		subjects, err := self.c.Git().Commit.GetShasAndCommitMessagesFirstLine(shas)
+		if err != nil {
+			return err
+		}
+		return self.c.ErrorMsg("Multiple base commits found. (" +
+			lo.Ternary(useIndex, "Try staging fewer changes", "Try staging some of the changes") +
+			")\n\n" + subjects)
+	}
+
+	commit, index, ok := lo.FindIndexOf(self.c.Model().Commits, func(commit *models.Commit) bool {
+		return commit.Sha == shas[0]
+	})
+	if !ok {
+		if self.c.Model().Commits[len(self.c.Model().Commits)-1].Status == models.StatusMerged {
```

Comment:
I would assign `self.c.Model().Commits` to a variable before this line so that if the slice gets swapped out concurrently we don't run the risk of an out-of-bounds exception

Reply:
This concurrency stuff still makes me dizzy... Like this? 092a40f803d1

Reply:
Yep that's right. To be clearer on what we're defending against, the original code is basically doing this:
```go
a := self.c.Model().Commits
b := self.c.Model().Commits
if a[len(b)-1].Status ...
```

So consider the following situation:
1) we evaluate `a` and it's a slice of 5 commits
2) before we evaluate b, a background refresh happens and our model commits slice is swapped out for a slice of 10 commits. Because it's swapped out with a brand new slice rather than extending the original slice, the slice that `a` points to is unchanged so `a` still has only 5 commits.
3) we evaluate b and that now has 10 commits because it's pointing to the new slice
4) we try to evaluate `a[9]` and get an out-of-bounds error because `a` only has 5 commits

Reply:
Thanks for detailed explanation. Thing is, I'm afraid we're still not making this thread-safe just by using a single variable. The assignment to the variable isn't atomic; a slice has three fields, a data pointer, a length, and a capacity. Our code might be preempted by a background refresh right after it has assigned the data pointer to the variable, but before it assigned the length to the variable. The only ways to solve this are 1. to use a mutex, or 2. to ensure that the data is only accessed from a single thread (which is this long-term project of mine that I keep talking about, and still haven't make much progress on).

For the purpose of this PR I wouldn't worry about it too much; we have tons of similar concurrency issues all over the code, and the likelihood of running into them in practice is low enough.